File: matlab_code/frames2signal.m
function [temporal_mean_filt, debug] = frames2signal(monoframes, conversion_method, fr, cutoff_freq)
	% Load constants
	constants;
	
	%=== Block 1. Convert the frame stream into a 1-D signal
	%conversion_method = 'simple-mean';
	switch conversion_method
		case 'simple-mean'
			temporal_mean = squeeze(nanmean(nanmean(monoframes, 1), 2));
			
		case 'trimmed-mean'	
			temporal_mean = squeeze(trimmean(reshape(monoframes, [size(monoframes, 1) * size(monoframes, 2), size(monoframes, 3)]), trimmed_size, 1));
		
		case 'mode-balance'
			% Selection parameters
			training_time = training_time_range; %seconds %Double
			lower_pct_range = pct_reach_below_mode; %Double
			upper_pct_range = pct_reach_above_mode; %Double
	
			% Find the mode of the pixel values in the first few frames
			stretched_first_frame = reshape(monoframes(:, :, round(fr * training_time(1)) + 1 : round(fr * training_time(2))), ...
											[size(monoframes, 1) * size(monoframes, 2) * (round(fr * training_time(2)) - round(fr * training_time(1))), 1]);
			%Double 1-D vector
			[counts, centres] = hist(stretched_first_frame, number_of_bins);
			%[Int vector, Double vector] 
			[~, argmax] = max(counts); %Int
			centre_mode = centres(argmax); %Double
			
			% Find the percentile range centred on the mode
			percentile_of_centre_mode = invprctile(stretched_first_frame, centre_mode); %Double
			percentile_range = [max(0, percentile_of_centre_mode - lower_pct_range), min(100, percentile_of_centre_mode + upper_pct_range)]; %Double 2-element vector
			
			% Correct the percentile range for the boundary cases
			if percentile_range(2) == 100
				percentile_range(1) = 100 - (lower_pct_range + upper_pct_range);
			end
			if percentile_range(1) == 0
				percentile_range(2) = (lower_pct_range + upper_pct_range);
			end
			
			% Convert the percentile range into pixel-value range
			range = prctile(stretched_first_frame, percentile_range); %Double 2-element vector
			
			% For each video frame, values outside the range are rejected
			monoframes((monoframes < range(1)) | (monoframes > range(2))) = NaN;
	
			% Calculate the average of each frame
			temporal_mean = squeeze(nanmean(nanmean(monoframes, 1), 2));
			
			debug.monoframes = monoframes;
	end
	
	%=== Block 2. Low-pass-filter the signal stream to remove unwanted noises
	filter_kernel = beatSignalFilterKernel;
    
    temporal_mean_filt = filter(filter_kernel, 1, temporal_mean);  %Double 1-D vector
	temporal_mean_filt = temporal_mean_filt((length(filter_kernel) - floor(length(filter_kernel) / 2)) : end);

File: MisfitHeartRate/MisfitHeartRate/algorithms/HeartRateCal/frames2signal.cpp
//
//  frames2signal.cpp
//  MisfitHeartRate
//
//  Created by Bao Nguyen on 7/2/14.
//  Copyright (c) 2014 misfit. All rights reserved.
//

#include "frames2signal.h"


namespace MHR {
    vector<double> frames2signal(const vector<Mat>& monoframes, String conversion_method,
                                 double fr, double cutoff_freq,
                                 double &lower_range, double &upper_range, bool isCalcMode,
                                 vector<Mat>& debug_monoframes)
    {
        clock_t t1 = clock();
        
        //=== Block 1. Convert the frame stream into a 1-D signal
        
        vector<double> temporal_mean;
        int height = monoframes[0].size.p[0];
        int width = monoframes[0].size.p[1];
        int total_frames = (int)monoframes.size();
        
        if(conversion_method == "simple-mean"){

            double size = height * width;
            for(int i=0; i<total_frames; ++i){
                double sum = 0;
                for(int x=0; x<height; ++x)
                    for(int y=0; y<width; ++y)
                        sum+=monoframes[i].at<double>(x,y);
                temporal_mean.push_back(sum/size);
            }
            
        }else if(conversion_method == "trimmed-mean"){
            
            //Set the trimmed size here
            int trimmed_size = _trimmed_size;
            
            double size = (height - trimmed_size * 2) * (width - trimmed_size * 2);
            for(int i=0; i<total_frames; ++i){
                double sum = 0;
                for(int x=trimmed_size; x<height-trimmed_size; ++x)
                    for(int y=trimmed_size; y<width-trimmed_size; ++y)
                        sum+=monoframes[i].at<double>(x,y);
                temporal_mean.push_back(sum/size);
            }
            
        }else if(conversion_method == "mode-balance"){
            
            // Selection parameters
//            double training_time = _training_time_end - _training_time_start;
            double lower_pct_range = _pct_reach_below_mode;
            double upper_pct_range = _pct_reach_above_mode;
            
            int first_tranning_frames_start = min( (int)round(fr * _training_time_start), total_frames );
            int first_tranning_frames_end = min( (int)round(fr * _training_time_end), total_frames );
//            int first_tranning_frames = min( (int)round(fr * training_time), total_frames );

            // this arr stores values of pixels from first trainning frames
            vector<double> arr;
            for(int i = first_tranning_frames_start; i < first_tranning_frames_end; ++i)
                for(int x=0; x<height; ++x)
                    for(int y=0; y<width; ++y)
                        arr.push_back(monoframes[i].at<double>(x,y));
            
            if (isCalcMode)
            {
                //find the mode
                vector<double> centres;
                vector<int> counts;
                
                hist(arr, _number_of_bins, counts, centres);
                
                int argmax=0;
                for(int i=0; i<(int)counts.size(); ++i) if(counts[i]>counts[argmax]) argmax = i;
                double centre_mode = centres[argmax];
                
                // find the percentile range centred on the mode
                double percentile_of_centre_mode = invprctile(arr, centre_mode);
                double percentile_lower_range = max(0.0, percentile_of_centre_mode - lower_pct_range);
                double percentile_upper_range = min(100.0, percentile_of_centre_mode + upper_pct_range);
                // correct the percentile range for the boundary cases
                if(percentile_upper_range == 100)
                    percentile_lower_range = 100 - (lower_pct_range + upper_pct_range);
                if(percentile_lower_range == 0)
                    percentile_upper_range = (lower_pct_range + upper_pct_range);
                
                //convert the percentile range into pixel-value range
                lower_range = prctile(arr, percentile_lower_range);
                upper_range = prctile(arr, percentile_upper_range);
            }
            
            printf("lower_range = %lf, upper_range = %lf\n", lower_range, upper_range);
            
            //now calc the avg of each frame while inogre the values outside the range
            double size = height * width;
            
            //this is the debug vector<Mat>
            debug_monoframes = vector<Mat>(monoframes);
            for(int i=0; i<total_frames; ++i){
                double sum = 0;
                int cnt = 0; //number of not-NaN-pixels
                for(int x=0; x<height; ++x)
                    for(int y=0; y<width; ++y){
                        double val=monoframes[i].at<double>(x,y);
                        if(val<lower_range - 1e-9 || val>upper_range + 1e-9){
                            val=0;
                            debug_monoframes[i].at<double>(x,y)=NaN; //set 0 for rejected pixels
                        }else ++cnt;
                        sum+=val;
                    }
                
                if(cnt==0) //push NaN for all-NaN-frames
                    temporal_mean.push_back(NaN);
                else
                    temporal_mean.push_back(sum/size);
            }
            
        }// end of mode-balance
        
        clock_t t2 = clock();
        printf("frames2signal() - Block 1 runtime = %f\n", ((float)t2 - (float)t1)/CLOCKS_PER_SEC);
        
        //=== Block 2. Low-pass-filter the signal stream to remove unwanted noises
        vector<double> temporal_mean_filt = low_pass_filter(temporal_mean);
        return temporal_mean_filt;
    }
}